src/main/java/com/hillel/items_exchange/service/impl/CategoryServiceImpl.java

```diff
@@ -0,0 +1,115 @@
+@RequiredArgsConstructor
+public class CategoryServiceImpl implements CategoryService {
+
+    private final CategoryRepository categoryRepository;
+    private final SubcategoryService subcategoryService;
+    private final ModelMapper modelMapper;
+
+    @Override
+    public List<String> findAllCategoryNames() {
+        return categoryRepository.findAllCategoriesNames();
+    }
+
+    @Override
+    public List<CategoryDto> findAllCategoryDto() {
+        return new ArrayList<>(convertAllTo(categoryRepository.findAll(),
+                CategoryDto.class, ArrayList::new));
+    }
+
+    @Override
+    public Optional<CategoryDto> findCategoryDtoById(long id) {
+        return categoryRepository.findById(id)
+                .map(category -> convertTo(category, CategoryDto.class));
+    }
+
+    @Override
+    public CategoryDto create(CategoryDto categoryDto) {
+        final Category savedCategory = saveCategory(categoryDto);
+        return convertTo(savedCategory, CategoryDto.class);
+    }
+
+    @Override
+    public CategoryDto update(CategoryDto categoryDto) {
+        final Category updatedCategory = saveCategory(categoryDto);
+        return convertTo(updatedCategory, CategoryDto.class);
+    }
+
+    @Override
+    public void removeById(long categoryId) {
+        categoryRepository.deleteById(categoryId);
+    }
+
+    @Override
+    public boolean isCategoryDtoDeletable(long categoryId) {
+        return isCategoryExistsById(categoryId) && isInternalSubcategoriesHaveNotProducts(categoryId);
+    }
+
+    @Override
+    public boolean isCategoryDtoUpdatable(CategoryDto categoryDto) {
+        return isCategoryExistsByIdAndNameOrNotExistsByName(categoryDto.getId(),
+                categoryDto.getName()) && isSubcategoryHasIdZeroOrIdExists(categoryDto);
+    }
+
+    @Override
+    public boolean isCategoryDtoValidForCreating(CategoryDto categoryDto) {
+        return isCategoryNameHasNotDuplicate(categoryDto.getName())
+                && categoryDto.getSubcategories().stream().allMatch(subcategoryDto -> subcategoryDto.getId() == 0L);
+    }
+
+    private boolean isCategoryExistsById(long categoryId) {
+        return categoryRepository.existsById(categoryId);
+    }
+
+    private boolean isCategoryNameHasNotDuplicate(String name) {
+        return !categoryRepository.existsByNameIgnoreCase(name);
+    }
+
+    private boolean isSubcategoryHasIdZeroOrIdExists(CategoryDto categoryDto) {
+        return categoryDto.getSubcategories().stream()
+                .filter(subcategoryDto -> subcategoryDto.getId() != 0)
+                .allMatch(subcategoryDto -> subcategoryService.isSubcategoryExistsById(subcategoryDto.getId()));
+    }
+
+    private boolean isInternalSubcategoriesHaveNotProducts(long categoryId) {
+        return categoryRepository.findById(categoryId)
+                .map(category -> category.getSubcategories().stream()
+                        .allMatch(subcategory -> subcategory.getProducts().isEmpty()))
```

Comment:
You could make it prettier: ``.map(Subcategory::getProducts).allMatch(Collection::isEmpty)``

Reply:
Cool!